MDC2020 Developments

AndrewEdmonds11 opened the pull request:
Hi Everyone

This PR combines PRs #231, #234 and #243. See those PRs for more information but briefly these add:

 -  mixing of StrawGasSteps (instead of tracker StepPointMCs)
 -  internal bookkeeping of simulation stage efficiencies
 - a new generator that represents all the products of stopped muons

and there is one backwards-breaking change:

 - cannot use MDC2018 background frames (i.e. dig.*-mix jobs)

I also have one question: do we want to update all the fcl in JobConfig to use the new features, or shall I leave that to the people running MDC2020 production?

Cheers
Andy

FNALbuild commented:
Hi @AndrewEdmonds11,
You have proposed changes to files in these packages:
- SimulationConfig
- TrackerMC
- Mu2eG4
- SimulationConditions
- DbTables
- JobConfig
- MCDataProducts
- Mu2eUtilities
- ProditionsService
- EventMixing
- EventGenerator
- GlobalConstantsService

which require these tests: build.

@Mu2e/write, @Mu2e/fnalbuild-users have access to CI actions on [master](/Mu2e/Offline/tree/master).

The following users requested to be notified about changes to these packages:
@resnegfk
:hourglass: The following tests have been triggered for baca00ca2fb57732cd27445d462b7096b47de0e3: build  (Build queue has 6 jobs)


[About FNALbuild.](https://mu2ewiki.fnal.gov/wiki/Git#GitHub_Pull_Request_Procedures_and_FNALbuild) [Code review on Mu2e/Offline.](https://mu2ewiki.fnal.gov/wiki/GitHubWorkflow#Code_Review)

FNALbuild commented:
:sunny: The tests passed at ref baca00ca2fb57732cd27445d462b7096b47de0e3.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| merge | :heavy_check_mark: | Merged baca00ca2fb57732cd27445d462b7096b47de0e3 at ae1f7c7601187452f44c1d9a636626b0d19c626c |
| scons build (prof) | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//248/artifact/scons.log). Build time: 12 min 15 sec |
| ceSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//248/artifact/ceSimReco.log) |
| g4test_03MT | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//248/artifact/g4test_03MT.log) |
| transportOnly | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//248/artifact/transportOnly.log) |
| PS | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//248/artifact/PS.log) |
| g4study | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//248/artifact/g4study.log) |
| cosmicSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//248/artifact/cosmicSimReco.log) |
| rootOverlaps | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//248/artifact/rootOverlaps.log) |
| g4surfaceCheck | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//248/artifact/g4surfaceCheck.log) |
| FIXME, TODO count | :wavy_dash: | [TODO (0) FIXME (1) in 23 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//248/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [10 errors 10 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//248/artifact/clang-tidy.log) |


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//248/console).
Build artefacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.


Logfiles may also be accessed [at this link.](https://gist.github.com/972f41dc80c51b8b0b737e9645e191f7)

brownd1978 commented:
Hi Andy, to answer your question: we need to update all the JobConfig fhicl.  The cleanest would be do to so as part of this PR, but it's not a requirement.  We can start the review without that.

FNALbuild commented:
:memo: The HEAD of `master` has changed to ded218ced0fec0e61a90d2521d5dd3419539ec9e. Tests are now out of date.

rlcee commented on SimulationConditions/inc/SimBookkeeperMaker.hh:
```diff
@@ -0,0 +1,43 @@
+  public:
+    SimBookkeeperMaker(SimBookkeeperConfig const& config):_config(config) {}
+
+    typename SimBookkeeper::ptr_t fromFcl() {
```
I don't recognize this - what is typename doing here?

AndrewEdmonds11 replied:
This looks like a copy error (I copied from TrkQualCatalogMaker which uses templates). I've removed "typename" from the file

rlcee commented on SimulationConditions/inc/SimBookkeeperMaker.hh:
```diff
@@ -0,0 +1,43 @@
+  public:
+    SimBookkeeperMaker(SimBookkeeperConfig const& config):_config(config) {}
+
+    typename SimBookkeeper::ptr_t fromFcl() {
```
I'd prefer to put these implementations in a cc file, just so that all proditions have the same file pattern

AndrewEdmonds11 replied:
Sorry, I saw this one after I committed. I'm splitting the implementations into a cc file and am trying to resolve some compilation errors. It should be done soon.

AndrewEdmonds11 replied:
This is now done :)

rlcee commented on SimulationConditions/inc/SimBookkeeper.hh:
```diff
@@ -0,0 +1,59 @@
+
+namespace mu2e {
+
+  class SimBookkeeper : virtual public ProditionsEntity {
```
I think using virtual here can mask a potential problem in the future.  I see some other ProditionsEntities use virtual and unless someone wants to defend it, I'll remove it there too.

AndrewEdmonds11 replied:
I've removed "virtual" from this file

rlcee reviewed (changes requested):
I reviewed the proditions part and only have a few small suggestions, so it looks good. 

One comment for everyone - I was surprised that this PR was made from 3 merged PR's.  I thought we wanted, in general, to make more single-topic bite-sized PR's

brownd1978 commented on DbTables/inc/SimEfficiencies.hh:
```diff
@@ -0,0 +1,79 @@
+    std::size_t nrow() const override { return _rows.size(); };
+    //    virtual std::size_t nrowFix() const { return 3; };
+    size_t size() const override { return _csv.capacity() +
+        + nrow()*nrow()/2 + nrow()*sizeof(Row); };
```
To answer your comment: we discussed merging with the author and decided it made most sense to combine them, due to overlapping developments.

brownd1978 replied:
But in general you are right: we should try to have PRs focus on one topic (I believe the wiki says that too).

AndrewEdmonds11 commented on DbTables/inc/SimEfficiencies.hh:
```diff
@@ -0,0 +1,79 @@
+      sstream << r.tag()<<",";
+      sstream << r.numerator()<<",";
+      sstream << r.denominator()<<",";
+      sstream << r.eff();
```
Added for the efficiency line. The numerator and denominator are integers so I hope those should be printed OK, but let me know if I should change those too.

rlcee commented on DbTables/inc/SimEfficiencies.hh:
```diff
@@ -41,8 +41,7 @@ namespace mu2e {
     //    virtual std::size_t nrowFix() const { return 3; };
-    size_t size() const override { return _csv.capacity() +
-        + nrow()*nrow()/2 + nrow()*sizeof(Row); };
+    size_t size() const override { return _csv.capacity(); };
```
Sorry, I meant to remove only the nrow*nrow part, so please restore  this: nrow()*sizeof(Row)

brownd1978 replied:
we have not adopted policy on resolving converstations.  I've used it to reduce the clutter on complicated feedback once an issue has been resolved, but of course that can be misleading if not everyone agrees it has been resolved.  For now I suggest either party may resolve a conversation if it's clear the issue is resolved.

AndrewEdmonds11 replied:
Whoops. I should have thought more about it. Added back.

rlcee replied:
It looks like you added it back on the 16th, but now it is gone again.

AndrewEdmonds11 replied:
I think it is still there (see https://github.com/Mu2e/Offline/pull/265/files#diff-f035c545608fa2fabc2ea0f195903e43bad3ba793f36f5a22358cb651345e015). It's just this comment thread attaches to a specific commit...

rlcee replied:
Clearly I have not mastered the review views..

rlcee reviewed (changes requested):
I note one more change, sorry I wasn't clear

I see there was no response to my comment on making simBookkeeperMaker.cc

Note to everyone.  Is there a way that I can see my comments and Andy's replies inline again?  I see them in the "conversation", but not in the "file changed".  Should I "resolve" the comment threads in the conversation?

AndrewEdmonds11 commented on EventGenerator/src/MuStopProductsGun_module.cc:
```diff
@@ -185,42 +171,45 @@ namespace mu2e {
+      // if (verbosityLevel_ > 0) {
+      //   std::cout << "MuStopProductsGun: Adding Capture Process: " << i_genPhys.pdgId << std::endl;
+      // }
+    //    }
```
I think that's already been deleted. At least, it is not in the version I see

AndrewEdmonds11 commented:
Before this gets merged in, I've just noticed that the straw hit rate from neutrons increases by 30% between NoPrimary-mix.MDC2020DEVa and the mu stop generator. The straw hit rate from photons also increases by 15%. The proton, deuteron and dio rates are all fine. I've been looking into it and will report when I work out what the issue is.

brownd1978 commented:
Could the neutron and photon rate changes be due to G4 physics list
changes?  MDC2020Dev is really MDC2018 so is 2 years old.

On Wed, Oct 14, 2020 at 9:15 AM Andrew Edmonds <notifications@github.com>
wrote:

> Before this gets merged in, I've just noticed that the straw hit rate from
> neutrons increases by 30% between NoPrimary-mix.MDC2020DEVa and the mu stop
> generator. The straw hit rate from photons also increases by 15%. The
> proton, deuteron and dio rates are all fine. I've been looking into it and
> will report when I work out what the issue is.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/265#issuecomment-708507875>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH572RIATJSDQA67L2LO3SKXFAZANCNFSM4SGVPAZQ>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

AndrewEdmonds11 commented:
It's certaintly possible - that's one of the things on my list to check

kutschke commented:
@FNALbuild run build tests

FNALbuild commented:
:hourglass: The following tests have been triggered for eb610d0dd31e9393dc086d018c410f607594fb26: build  (Build queue is empty)

FNALbuild commented:
:bangbang: The build test could not run due to merge conflicts. Please resolve these first and try again.
```
> git diff --check | grep -i conflict
CommonMC/fcl/prolog.fcl:14: leftover conflict marker
CommonMC/fcl/prolog.fcl:20: leftover conflict marker
CommonMC/fcl/prolog.fcl:31: leftover conflict marker
```

AndrewEdmonds11 commented:
Looking into the increase in straw hit rates:
- I don't think it's a geometry difference
    - the only difference between the two geometries (from what I could tell) were:
        - muonBeamStop v04 --> v05
        - ProductionTargetInPS --> ProductionTarget_Hayman_v2_0
    - I can't see any differences in the location of particles leading to straw hits with shdiag
- could be a geant4 difference
    - MDC2020Dev was a reprocessing of MDC2018 (Offline v7_5_1), which used Geant4.10.4.p03b and the new code is using Geant4.10.6.p02b
    - it was noted in the upgrade to Geant4.10.6.p02b that we see more low-energy electrons and photons, and the low-energy behaviour of neutrons changed in the POT simulation (DocDB-32034)
    - I've not done a direct comparison with different geant4 versions (if anyone knows how to do this, then let me know)

Is there anything else I should check?

Going forward, I tihnk we should consider adding the new MuStopGenerator with digi making to the validation. Then we can plot our standard straw hit times plot (without the flash) and keep track of when these changes happen.

kutschke commented:
@FNALbuild run build tests

FNALbuild commented:
:hourglass: The following tests have been triggered for a575d40d05fa493faa2ab0ae2c99e950c2206fa9: build  (Build queue has 2 jobs)

FNALbuild commented:
:umbrella: The tests failed for ref a575d40d05fa493faa2ab0ae2c99e950c2206fa9.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| merge | :heavy_check_mark: | Merged a575d40d05fa493faa2ab0ae2c99e950c2206fa9 at ded218ced0fec0e61a90d2521d5dd3419539ec9e |
| scons build (prof) | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//257/artifact/scons.log). Build time: 14 min 21 sec |
| ceSimReco | :x: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//257/artifact/ceSimReco.log) |
| g4test_03MT | :x: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//257/artifact/g4test_03MT.log) |
| transportOnly | :x: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//257/artifact/transportOnly.log) |
| PS | :x: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//257/artifact/PS.log) |
| g4study | :x: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//257/artifact/g4study.log) |
| cosmicSimReco | :x: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//257/artifact/cosmicSimReco.log) |
| rootOverlaps | :x: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//257/artifact/rootOverlaps.log) |
| g4surfaceCheck | :x: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//257/artifact/g4surfaceCheck.log) |
| FIXME, TODO count | :wavy_dash: | [TODO (0) FIXME (1) in 27 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//257/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [10 errors 10 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//257/artifact/clang-tidy.log) |


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//257/console).
Build artefacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.


Logfiles may also be accessed [at this link.](https://gist.github.com/ee9f5a9b636738cd40eb661f1864a502)

AndrewEdmonds11 commented:
Whoops. Looks like I added an extra ']' in CommonMC/fcl/prolog.fcl. Fixing now and will double-check ceSimReco.fcl locally before committing

kutschke commented:
@FNALbuild run build tests

FNALbuild commented:
:hourglass: The following tests have been triggered for 148860168cc1ab657ffe63e9192276486122dc6e: build  (Build queue has 3 jobs)

FNALbuild commented:
:sunny: The tests passed at ref 148860168cc1ab657ffe63e9192276486122dc6e.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| merge | :heavy_check_mark: | Merged 148860168cc1ab657ffe63e9192276486122dc6e at ded218ced0fec0e61a90d2521d5dd3419539ec9e |
| scons build (prof) | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//258/artifact/scons.log). Build time: 13 min 41 sec |
| ceSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//258/artifact/ceSimReco.log) |
| g4test_03MT | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//258/artifact/g4test_03MT.log) |
| transportOnly | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//258/artifact/transportOnly.log) |
| PS | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//258/artifact/PS.log) |
| g4study | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//258/artifact/g4study.log) |
| cosmicSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//258/artifact/cosmicSimReco.log) |
| rootOverlaps | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//258/artifact/rootOverlaps.log) |
| g4surfaceCheck | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//258/artifact/g4surfaceCheck.log) |
| FIXME, TODO count | :wavy_dash: | [TODO (0) FIXME (1) in 27 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//258/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [10 errors 10 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//258/artifact/clang-tidy.log) |


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//258/console).
Build artefacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.


Logfiles may also be accessed [at this link.](https://gist.github.com/b24d1f1e6b0b647213e25f0996028585)

gaponenko commented:
> Looking into the increase in straw hit rates:  ...   could be a geant4 difference

Hi Andy,  I suggest you compare the rates between your "final" version with all the commits, and the "base" version of your branch.  Rather than comparing to some older existing sample.

FNALbuild commented:
:memo: The HEAD of `master` has changed to c3344ca9e93c48d678a2352daf95917c457cf107. Tests are now out of date.

gaponenko commented on EventGenerator/src/DIOGenerator_tool.cc:
```diff
@@ -0,0 +1,68 @@
+    double _mass;
+    GenId _genId;
+
+    SpectrumVar       _spectrumVariable;
```
Suggest to remove _spectrumVariable.  This is a DIO generator, isn't it always going to be energy? (Unless someone make a mistake in  fhicl config.)

AndrewEdmonds11 replied:
Done in commit 1c746446b

gaponenko commented on EventGenerator/src/MuCapGammaRayGenerator_tool.cc:
```diff
@@ -0,0 +1,65 @@
+    const CLHEP::Hep3Vector pos(stop.x, stop.y, stop.z);
+
+    double rand = _randFlat->fire();
+    if (rand < _intensity) {
```
This code produces mean number of gammas per capture == intensity <= 1.     In our past simulations we assumed 2 gammas per capture.  I think the code has to change to sample the number of gammas from a Poisson distribution.

AndrewEdmonds11 replied:
This generator is for the 1.8 MeV gamma ray and is not given to the MuStopGenerator. It is here as an additional example that might be useful. The tool that we will use is MuCapPhotonGenerator and does the Poisson distribution with the assumed 2 photons per capture. I will rename this generator tool to MuCap1809keVGammaRayGenerator_tool or something similar to avoid confusion.

gaponenko commented on EventGenerator/src/MuCapGammaRayGenerator_tool.cc:
```diff
@@ -0,0 +1,65 @@
+    double _energy;
+    double _intensity;
+
+    SpectrumVar       _spectrumVariable;
+    BinnedSpectrum    _spectrum;
```
Please remove unused _spectrumVariable and _spectrum.    It would also make sense to do away wiith _mass for the photon generator.

AndrewEdmonds11 replied:
Done in commit 6af04d720 (along with generator tool renaming)

AndrewEdmonds11 commented:
Following up on the different straw hit rates from neutrons between MDC2020DEVa and the MuStopGenerator: I created new neutron files containing StepPointMCs with this branch. I moved StrawGasStep creation back to the digi stage and for mixing used the old "meanEventsPerProton" method rather than any of the new stuff. I recalculated this value for the new neutron file and it came out to be 2.138420880E-05 (for MDC2018 it was 1.8790E-05). The straw hit rates now match so the new generator is validated.

gaponenko commented on EventGenerator/src/MuCapPhotonGenerator_tool.cc:
```diff
@@ -0,0 +1,76 @@
+
+  private:
+    PDGCode::type _pdgId;
+    double _mass;
+    GenId _genId;
+    double _rate;
+
+    SpectrumVar       _spectrumVariable;
```
MuCapPhotonGenererator does not need _mass and _spectrumVariable; please simplify.

AndrewEdmonds11 replied:
Done in commit ca938c135

gaponenko commented on EventGenerator/src/SConscript:
```diff
@@ -51,6 +51,7 @@ mainlib = helper.make_mainlib ( [
         'HepPDT',
         'HepPID',
         rootlibs,
+        'boost_filesystem',
```
I do not see any uses of boost_filesystem in EventGenerators.  Why was the link dependency added?

AndrewEdmonds11 replied:
Whoops. I must have been playing around trying to solve a compilation error. Removed in commit a4ce17523

kutschke commented:
Somewhere in my notes I have the prescription to tell gcc to report unused link libraries.   We should scrub the link lists as targets of opportunity.

On 10/26/20, 10:33 AM, "Andrei Gaponenko" <notifications@github.com<mailto:notifications@github.com>> wrote:


@gaponenko commented on this pull request.

________________________________

In EventGenerator/src/SConscript<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_265-23discussion-5Fr512053876&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=PQrdT681Qsp8AtiUe5Tq5GOV7czS2eGaXH0pSlve3mk&s=ltBBQZTU_UlDjt6lhNJj-W-SGqlTuxfyFJago7EQdMQ&e=>:

> @@ -51,6 +51,7 @@ mainlib = helper.make_mainlib ( [

         'HepPDT',

         'HepPID',

         rootlibs,

+        'boost_filesystem',

I do not see any uses of boost_filesystem in EventGenerators. Why was the link dependency added?

—
You are receiving this because you are on a team that was mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_265-23pullrequestreview-2D516901637&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=PQrdT681Qsp8AtiUe5Tq5GOV7czS2eGaXH0pSlve3mk&s=R1KJeMXbmdxKn_ic7QBcdMdWOVN0fz9JOV92Twvd0Fo&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABHY5ZVUBFJ26SAK4I7Q4FDSMWJCXANCNFSM4SGVPAZQ&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=PQrdT681Qsp8AtiUe5Tq5GOV7czS2eGaXH0pSlve3mk&s=ouZv-H0AUj9wLnnJX-07Z7Rs1kb2jh7NvrsOtpkSva0&e=>.

gaponenko commented on EventMixing/src/MixBackgroundFrames_module.cc:
```diff
@@ -121,28 +137,58 @@ namespace mu2e {
     if(writeEventIDs_) {
       helper.produces<art::EventIDSequence>();
     }
+    if (pars().mu2e().meanEventsPerProton(meanEventsPerProton_)) {
+      mixingMeanOverride_ = true;
+    }
```
Please disallow specifying both meanEventsPerProton and new style configs.   Inside this if(),  throw an exeception if either of simStageEfficiencyTags or meanEventsPerPOTFactors not empty.

AndrewEdmonds11 replied:
Done (commit c6d970e47)

gaponenko commented on EventMixing/src/MixBackgroundFrames_module.cc:
```diff
@@ -121,28 +137,58 @@ namespace mu2e {
     pbi_ = *event.getValidHandle<ProtonBunchIntensity>(pbiTag_);
     if(debugLevel_ > 0)std::cout << " Starting event mixing, Intensity = " << pbi_.intensity() << std::endl;
+
+    SimBookkeeper const& simbookkeeper = _simbookkeeperH.get(event.id());
```
Talking to the DB should be conditional.  If one specified meanEventsPerProton no handle should be created here.

AndrewEdmonds11 replied:
Done (commit c6d970e47)

gaponenko commented on JobConfig/beam/prolog.fcl:
```diff
@@ -217,7 +218,9 @@ mu2e.physics.g4s4Muons.producers.generate: {
    physics: @nil
-   muonStops : @local::mu2e.tgtMuonStops
+   muonStops : { @table::mu2e.tgtMuonStops
+   	       	 inputFiles : @local::mergedMuon_tgtStops_mdc2018
```
This replaces @nil with a hardcoded set of input files.   I think this change is outside the scope of this PR, please take it out.

AndrewEdmonds11 replied:
Done (commit 3e76dc9e4)

gaponenko commented on JobConfig/beam/oot.fcl:
```diff
@@ -18,7 +18,9 @@ physics.producers.generate: {
     pdgId: 13 // mu-
-    muonStops : @local::mu2e.ootMuonStops
+    muonStops : { @table::mu2e.ootMuonStops
+   	       	  inputFiles : @local::mergedMuon_ootStops_mdc2018
```
Same for OOT: this replaces @nil with a hardcoded set of input files.   I think this change is outside the scope of this PR, please take it out.

AndrewEdmonds11 replied:
Done (commit 3e76dc9e4)

gaponenko commented on GlobalConstantsService/inc/PhysicsParams.hh:
```diff
@@ -109,6 +109,32 @@ namespace mu2e
+      const std::string allowedMaterial = checkMaterial( material );
+      return _captureNeutronRate.find(allowedMaterial)->second;
+    }
+    double   getCapturePhotonRate     (targetMat material = "") const {
+      const std::string allowedMaterial = checkMaterial( material );
+      return _capturePhotonRate.find(allowedMaterial)->second;
+    }
+
+    double   getCaptureGammaEnergy     (targetMat material = "") const {
+      const std::string allowedMaterial = checkMaterial( material );
+      return _captureGammaEnergy.find(allowedMaterial)->second;
+    }
+    double   getCaptureGammaIntensity     (targetMat material = "") const {
+      const std::string allowedMaterial = checkMaterial( material );
+      return _captureGammaIntensity.find(allowedMaterial)->second;
+    }
```
It is confusing to have both "capture photons" and "capture gammas".  With different parameterizations (rate vs intensity) on top of that.   Do we REALLY need both?

AndrewEdmonds11 replied:
I will finish renaming the capture gamma to 1809keVGamma to match the renaming of the generator tool. Hopefully this will clear some confusion.

To your more fundamental question: I think there will be a use for both. I believe the calorimeter plans to see if they can detect the 1809 keV gamma and knowing its intensity is necessary to normalize. Being able to easily simulate this on top of our flat photon spectrum will be useful.

I'm happy to remove it if people want me to.

AndrewEdmonds11 replied:
Renaming finished in commit 1f23a940f

kutschke commented:
Just curious:  is the intention to distinguish atomic transition vs nuclear transition?  If so, I think that standard usage is x-ray (atomic) and gamma-ray( nuclear), with photon being a collective noun.  Or are you inheriting language from another source (G4?)?


On 10/26/20, 11:47 AM, "Andrei Gaponenko" <notifications@github.com<mailto:notifications@github.com>> wrote:


@gaponenko commented on this pull request.

________________________________

In GlobalConstantsService/inc/PhysicsParams.hh<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_265-23discussion-5Fr512109617&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=3jTT2fbIMIwmusv63IuXgzSQZdlqDuAzRIfS_j1gESU&s=dcbW163UgKZLkLBd9_NK7dMwmOU0g-3rmmId7raUrAc&e=>:

> +    double   getCapturePhotonRate     (targetMat material = "") const {

+      const std::string allowedMaterial = checkMaterial( material );

+      return _capturePhotonRate.find(allowedMaterial)->second;

+    }

+

+    double   getCaptureGammaEnergy     (targetMat material = "") const {

+      const std::string allowedMaterial = checkMaterial( material );

+      return _captureGammaEnergy.find(allowedMaterial)->second;

+    }

+    double   getCaptureGammaIntensity     (targetMat material = "") const {

+      const std::string allowedMaterial = checkMaterial( material );

+      return _captureGammaIntensity.find(allowedMaterial)->second;

+    }

It is confusing to have both "capture photons" and "capture gammas". With different parameterizations (rate vs intensity) on top of that. Do we REALLY need both?

—
You are receiving this because you are on a team that was mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_265-23pullrequestreview-2D516977550&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=3jTT2fbIMIwmusv63IuXgzSQZdlqDuAzRIfS_j1gESU&s=C-V72zhwDXdaFmEvJrGZ1Vsu_aCnUfeYztKXTZS-MJw&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABHY5ZVF45J567RTFKCMIE3SMWR2DANCNFSM4SGVPAZQ&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=3jTT2fbIMIwmusv63IuXgzSQZdlqDuAzRIfS_j1gESU&s=IL5CUyeI-njpjqIHae8w5akW4hm6AnOkqts5l0mnJr0&e=>.

gaponenko commented:
Atomic cascade was a part of an earlier PR that got bundled into the 
current one; I think it has been taken out per my request.  The current 
code has two kinds of nuclear capture photons.

Andrei

On Mon, 26 Oct 2020, Rob Kutschke wrote:

> 
>   Just curious:  is the intention to distinguish atomic transition vs nuclear transition?  If so, I think that standard usage is x-ray (atomic) and gamma-ray( nuclear), with photon being a collective noun.  Or are you inheriting language from another source (G4?)?
> 
> 
> On 10/26/20, 11:47 AM, "Andrei Gaponenko" <notifications@github.com<mailto:notifications@github.com>> wrote:
> 
> 
> @gaponenko commented on this pull request.
> 
> ________________________________
> 
> In GlobalConstantsService/inc/PhysicsParams.hh<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_265-23discussion-5Fr512109617&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=3jTT2fbIMIwmusv63IuXgzSQZdlqDuAzRIfS_j1gESU&s=dcbW163UgKZLkLBd9_NK7dMwmOU0g-3rmmId7raUrAc&e=>:
> 
> > +    double   getCapturePhotonRate     (targetMat material = "") const {
> 
> +      const std::string allowedMaterial = checkMaterial( material );
> 
> +      return _capturePhotonRate.find(allowedMaterial)->second;
> 
> +    }
> 
> +
> 
> +    double   getCaptureGammaEnergy     (targetMat material = "") const {
> 
> +      const std::string allowedMaterial = checkMaterial( material );
> 
> +      return _captureGammaEnergy.find(allowedMaterial)->second;
> 
> +    }
> 
> +    double   getCaptureGammaIntensity     (targetMat material = "") const {
> 
> +      const std::string allowedMaterial = checkMaterial( material );
> 
> +      return _captureGammaIntensity.find(allowedMaterial)->second;
> 
> +    }
> 
> It is confusing to have both "capture photons" and "capture gammas". With different parameterizations (rate vs intensity) on top of that. Do we REALLY need both?
> 
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_265-23pullrequestreview-2D516977550&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=3jTT2fbIMIwmusv63IuXgzSQZdlqDuAzRIfS_j1gESU&s=C-V72zhwDXdaFmEvJrGZ1Vsu_aCnUfeYztKXTZS-MJw&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABHY5ZVF45J567RTFKCMIE3SMWR2DANCNFSM4SGVPAZQ&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=3jTT2fbIMIwmusv63IuXgzSQZdlqDuAzRIfS_j1gESU&s=IL5CUyeI-njpjqIHae8w5akW4hm6AnOkqts5l0mnJr0&e=>.
> 
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_265-23issuecomment-2D716681800&d=DwIFaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=O47fc5vzDTR2V_gla4Ub0Q&m=8-FPbrYTi_jAJ-UqJD-RVcjXuo6Zg5doPCDKn7dztPI&s=aPNZ2_bOMQoXX6NJRxEyQ5TCFzys73VcSJSIDAC2Ne0&e=

brownd1978 commented:
What is the timing of this photon WRT the other muon capture products?  If
it is synchronous it belongs in this generator, if not, not.

On Mon, Oct 26, 2020 at 10:04 AM Andrew Edmonds <notifications@github.com>
wrote:

> *@AndrewEdmonds11* commented on this pull request.
> ------------------------------
>
> In GlobalConstantsService/inc/PhysicsParams.hh
> <https://github.com/Mu2e/Offline/pull/265#discussion_r512125443>:
>
> > +    double   getCapturePhotonRate     (targetMat material = "") const {
> +      const std::string allowedMaterial = checkMaterial( material );
> +      return _capturePhotonRate.find(allowedMaterial)->second;
> +    }
> +
> +    double   getCaptureGammaEnergy     (targetMat material = "") const {
> +      const std::string allowedMaterial = checkMaterial( material );
> +      return _captureGammaEnergy.find(allowedMaterial)->second;
> +    }
> +    double   getCaptureGammaIntensity     (targetMat material = "") const {
> +      const std::string allowedMaterial = checkMaterial( material );
> +      return _captureGammaIntensity.find(allowedMaterial)->second;
> +    }
>
> I will finish renaming the capture gamma to 1809keVGamma to match the
> renaming of the generator tool. Hopefully this will clear some confusion.
>
> To your more fundamental question: I think there will be a use for both. I
> believe the calorimeter plans to see if they can detect the 1809 keV gamma
> and knowing its intensity is necessary to normalize. Being able to easily
> simulate this on top of our flat photon spectrum will be useful.
>
> I'm happy to remove it if people want me to.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/265#discussion_r512125443>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH57YJEFKCQMYFYU2WGRDSMWT3RANCNFSM4SGVPAZQ>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

AndrewEdmonds11 commented:
The 1809 keV gamma is coincident with the muon capture. However, something that I think is being missed in this discussion is that it is not currently included in the fcl configuration for the MuStopGenerator. I have written it as an additional example of writing a generator tool and as something that might be useful later.

gaponenko commented:
> To your more fundamental question: I think there will be a use for both. I believe the calorimeter plans to see if they can detect the 1809 keV gamma and knowing its intensity is necessary to normalize. Being able to easily simulate this on top of our flat photon spectrum will be useful.
> 
> I'm happy to remove it if people want me to.

It sounds like this should stay.  It is a refinement of our existing nuclear capture simulation.
I think the renaming will help.

Andrei

> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_265-23discussion-5Fr512125443&d=DwICaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=O47fc5vzDTR2V_gla4Ub0Q&m=mAB4XbPVJR1pw19bgsiUkRcbRWgec3Tmq1eD1QQ752M&s=yYdxb7sOH4iw8J20SuVNRyVjvSGc2MtaMg8JsN3kHU8&e=

gaponenko commented on SimulationConditions/inc/SimBookkeeper.hh:
```diff
@@ -0,0 +1,48 @@
+#define SimulationConditions_SimBookkeeper_hh
+
+//
+// SimBookkeeper is the ProditionsEntry for
```
Incomplete comment?

AndrewEdmonds11 replied:
Completed in commit 354fb8ff2

gaponenko commented on SimulationConditions/src/SimBookkeeper.cc:
```diff
@@ -0,0 +1,23 @@
+//
+// SimBookkeeper is the ProditionsEntry for
```
Another incomplete comment.   By the way,   if the purpose of a class is explained in the header, I don't think it needs to be repeated in the implementation file.

AndrewEdmonds11 replied:
Removed since I added it in the header (commit 354fb8ff2)

brownd1978 commented:
If it's just a tool I see no reason to remove it.  Same would be true for
atomic capture gammas.
Dave

On Mon, Oct 26, 2020 at 10:23 AM Andrew Edmonds <notifications@github.com>
wrote:

> The 1809 keV gamma is coincident with the muon capture. However, something
> that I think is being missed in this discussion is that it is not currently
> included in the fcl configuration for the MuStopGenerator. I have written
> it as an additional example of writing a generator tool and as something
> that might be useful later.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/265#issuecomment-716702900>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH573MKD52VCIP35VHTXDSMWWCBANCNFSM4SGVPAZQ>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

gaponenko commented on SimulationConfig/inc/SimBookkeeperConfig.hh:
```diff
@@ -0,0 +1,35 @@
+    fhicl::Atom<std::string> tag{
+      Name("tag"), Comment("Identifying tag for this efficiency")};
+    fhicl::Atom<double> eff{
+      Name("eff"), Comment("EFficiency value")};
```
It looks like one must specify  stage efficiencies even when the number will be read from a database?
May be I do not understand the intent of the current interface.     I think that the required configuration should contain just a set of "tags".  Then efficiency values for the tags can either be read from the database, or provided via a separate piece of FHICL (which must be absent if DB is being used).

AndrewEdmonds11 replied:
I think this is the "separate piece of FHICL" that can be used to override what's in the database. I'm not sure if it has to be absent if the database is being used...

gaponenko commented on TrackerMC/fcl/prolog.fcl:
```diff
@@ -8,7 +8,7 @@ makeSGS : {
     module_type : StrawDigisFromStrawGasSteps
     TimeOffsets   : [  @sequence::CommonMC.TimeMaps ]
-    StrawGasStepModule : makeSGS
+    StrawGasStepModule : ""
```
I do not understand this change, can you please explain?

AndrewEdmonds11 replied:
Before we stored StepPointMCs in each background frame and they were all read in a single makeSGS module to create a single StrawGasStepCollection. Now we store StrawGasSteps in each background frame and so makeSD needs to do the combining. This functionality already exists in makeSD and we supply an empty input tag to do it.

gaponenko commented:
Thanks, Andy!
Andrei

On Mon, 26 Oct 2020, Andrew Edmonds wrote:

> @AndrewEdmonds11 commented on this pull request.
> 
> 
> 
> > @@ -8,7 +8,7 @@ makeSGS : {
>  makeSD : {
>      module_type : StrawDigisFromStrawGasSteps
>      TimeOffsets   : [  @sequence::CommonMC.TimeMaps ]
> -    StrawGasStepModule : makeSGS
> +    StrawGasStepModule : ""
> 
> Before we stored StepPointMCs in each background frame and they were all read in a single makeSGS module to create a single StrawGasStepCollection. Now we store StrawGasSteps in each background frame and so makeSD needs to do the combining. This functionality already exists in makeSD and we supply an empty input tag to do it.
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_265-23discussion-5Fr512152893&d=DwICaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=O47fc5vzDTR2V_gla4Ub0Q&m=aZnKslroPEe0cZ4UIHCttHmaECdHjvGUZPsuwTXZV7k&s=TCIxShcZ5hcRDRzRjKl3ckt4Z7jkdZ53V0yxPCC7vqU&e=

gaponenko commented:
> I think this is the "separate piece of FHICL" that can be used to override what's in the database.

OK, it makes sense then.
Thank you,
Andrei

gaponenko commented on EventGenerator/src/MuStopProductsGun_module.cc:
```diff
@@ -0,0 +1,156 @@
+    int               verbosityLevel_;
+
+    art::RandomNumberGenerator::base_engine_t& eng_;
+    RandomUnitSphere   randomUnitSphere_;
```
Not needed

AndrewEdmonds11 replied:
Removed in commit 27d481d68

gaponenko commented on EventGenerator/src/MuStopProductsGun_module.cc:
```diff
@@ -0,0 +1,156 @@
+#include "DataProducts/inc/PDGCode.hh"
+#include "MCDataProducts/inc/GenParticle.hh"
+#include "MCDataProducts/inc/GenParticleCollection.hh"
+#include "Mu2eUtilities/inc/RandomUnitSphere.hh"
+#include "Mu2eUtilities/inc/CzarneckiSpectrum.hh"
+#include "Mu2eUtilities/inc/ConversionSpectrum.hh"
+#include "Mu2eUtilities/inc/SimpleSpectrum.hh"
+#include "Mu2eUtilities/inc/EjectedProtonSpectrum.hh"
+#include "Mu2eUtilities/inc/BinnedSpectrum.hh"
+#include "Mu2eUtilities/inc/Table.hh"
```
Please remove the leftover includes.  Also RandPoissonQ.h

AndrewEdmonds11 replied:
Done in commit 27d481d68

gaponenko commented on JobConfig/beam/mustop.fcl:
```diff
@@ -0,0 +1,92 @@
+outputs.crvOutput.fileName: "sim.owner.mustop-CRV.version.sequencer.art"
+outputs.truncatedEvtsOutput.fileName: "sim.owner.mustop-truncated.version.sequencer.art"
+
+physics.producers.generate.stops.inputFiles : [ "mergedMuonStops/nts.mu2e.DS-TGTstops.MDC2018a.001002_00000000.root" ]
```
That's another hardcoded input file.  I think production job configs should have @nil for that.

AndrewEdmonds11 replied:
Done in commit 17349b7c2

gaponenko reviewed (changes requested):
Andy, thank you for resolving all the line level code requests.  It
looks like the stuff from former PRs #234 and #243 could be merged
now.  Unfortunately it got bundled up with #231, and I want more
changes to that part.

MuStopProductsGun is a huge improvements compared to the initial
version in the previous PR.  The code is easy to read, and it clearly
says what is does.  Using the tools helped.  What is left to do:
configuration clean up.  I feel strongly that GenPhysConfig should go.
As I mentioned in my review for #231, some of the issues stem from the
existing BinnedSpectrum code.  We can kick that can down the road and
not touch BinnedSpectrum for this PR, but at least we should not make
things worse.   Specifically, my suggestion is:

- completely remove the GenPhysConfig.hh file, and related changes in
  BinnedSpectrum.

- ParticleGeneratorTool.hh does not need to defined Config, take it
  out.  Instead each concrete tool class should define its own Config
  with just the fields that are needed for that generator.  Some of
  them need BinnedSpectrum but not mass, others need mass, etc.
  A working code example is in
  /mu2e/app/users/gandr/Offline.andy/EventGenerator/src/MuCapPhotonGenerator_tool.cc

- At this point you can reduce the nestedness of fhicl config for the
  tools.  The original StoppedParticleReactionGun defined the
  "physics" subset to separate it from the muon stop sampling
  configuration and other top level stuff.   For the tools
  there is no other stuff, so you can as well get rid of the "physics"
  level that is left in my test muonCapturePhotonGenTool config in
  /mu2e/app/users/gandr/Offline.andy/JobConfig/beam/mustop.fcl

Also, please clean up white spaces issues. For example, there are both
hard tabs and trailing spaces in mustop.fcl
(I thought that would be caught by a CI script these days???)

Andrei

AndrewEdmonds11 commented:
Thanks for the review, Andrei. I've done that configuration simplification in commit 00615dc, and resolved whitespace issues in the files I touched in commit a7f0379cb

gaponenko commented on Mu2eUtilities/inc/BinnedSpectrum.hh:
```diff
@@ -50,6 +52,7 @@ namespace mu2e {
     double         getXMaxUnbinned() const { return _xmax_unbinned;}
     double         getXMax() const { return _xmax;}
     double         getXMin() const { return _xmin;}
+    std::string    spectrumVariable() const { return _spectrumVariable;}
```
This is still adding extra stuff to BinnedSpectrum.   Which BinnedSpectrum has no use for.  Please put spectrumVariable  configuration instead into tools that need it, like EventGenerator/src/MuCapNeutronGenerator_tool.cc   I think you do not need to modify BinnedSpectrum from its original version.   The fhicl config will look like

   spectrumVariable: kineticEnergy
   spectrum: { ... }

instead of having spectrumVariable inside; I think this is fine.

AndrewEdmonds11 replied:
Thanks, I was having issues getting spectrumVariable out of spectrum so put it into BinnedSpectrum as that seemed like the simplest solution. I will move it into the tool config like you suggest.

AndrewEdmonds11 replied:
Done in commit 429a1beba

gaponenko replied:
> Done in commit [429a1be](https://github.com/Mu2e/Offline/commit/429a1beba1ff881af84ed52bbe659df05bd0928c)

Thank you!

kutschke commented:
Please give me a heads up when you expect a period of inactivity.  I will then launch another round of tests.

gaponenko reviewed (approved):
Andy, thank you  for being so responsive! 
Rob, please take over and run the tests.
Andrei

kutschke commented:
@FNALbuild run build tests

FNALbuild commented:
:hourglass: The following tests have been triggered for 429a1beba1ff881af84ed52bbe659df05bd0928c: build  (Build queue has 2 jobs)

kutschke commented:
Andy and Ray: according to this page there is a comment from Ray outstanding.  Has it in fact been resolved?

FNALbuild commented:
:sunny: The tests passed at ref 429a1beba1ff881af84ed52bbe659df05bd0928c.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| merge | :heavy_check_mark: | Merged 429a1beba1ff881af84ed52bbe659df05bd0928c at c3344ca9e93c48d678a2352daf95917c457cf107 |
| scons build (prof) | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//263/artifact/scons.log). Build time: 13 min 23 sec |
| ceSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//263/artifact/ceSimReco.log) |
| g4test_03MT | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//263/artifact/g4test_03MT.log) |
| transportOnly | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//263/artifact/transportOnly.log) |
| PS | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//263/artifact/PS.log) |
| g4study | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//263/artifact/g4study.log) |
| cosmicSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//263/artifact/cosmicSimReco.log) |
| rootOverlaps | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//263/artifact/rootOverlaps.log) |
| g4surfaceCheck | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//263/artifact/g4surfaceCheck.log) |
| FIXME, TODO count | :wavy_dash: | [TODO (0) FIXME (1) in 25 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//263/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [10 errors 7 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//263/artifact/clang-tidy.log) |


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//263/console).
Build artefacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.


Logfiles may also be accessed [at this link.](https://gist.github.com/0dc6d7e16572a1fc0c9bff6718e0cd67)

AndrewEdmonds11 commented:
Hi Rob. I think that comment has been resolved. I added .cc files for the SimBookkeeper classes and corrected my nrow()*sizeof(Row) mistake. I'll let Ray confirm or not.
